plugin/src/main/java/org/autorefactor/refactoring/rules/MapEliminateKeySetCallsRefactoring.java

```diff
@@ -50,39 +63,140 @@ public String getName() {
-                    removeInvocationOfMapKeySet(parentMi, mi, "containsKey");
-                    return DO_NOT_VISIT_SUBTREE;
-                }
+            if (isMethod(mi, "java.util.Set", "clear")) {
+                removeInvocationOfMapKeySet(parentMi, mi, "clear");
+                return DO_NOT_VISIT_SUBTREE;
+            }
+            if (isMethod(mi, "java.util.Set", "size")) {
+                removeInvocationOfMapKeySet(parentMi, mi, "size");
+                return DO_NOT_VISIT_SUBTREE;
+            }
+            if (isMethod(mi, "java.util.Set", "isEmpty")) {
+                removeInvocationOfMapKeySet(parentMi, mi, "isEmpty");
+                return DO_NOT_VISIT_SUBTREE;
+            }
+            if (isMethod(mi, "java.util.Set", "remove", "java.lang.Object")) {
+                removeInvocationOfMapKeySet(parentMi, mi, "remove");
+                return DO_NOT_VISIT_SUBTREE;
+            }
+            if (isMethod(mi, "java.util.Set", "contains", "java.lang.Object")) {
+                removeInvocationOfMapKeySet(parentMi, mi, "containsKey");
+                return DO_NOT_VISIT_SUBTREE;
             }
         }
         return VISIT_SUBTREE;
     }
 
-    private void removeInvocationOfMapKeySet(
-            MethodInvocation mapKeySetMi, MethodInvocation actualMi, String methodName) {
+    private boolean isKeySetMethod(Expression expr) {
+        if (expr instanceof MethodInvocation) {
+            return isMethod((MethodInvocation) expr, "java.util.Map", "keySet");
+        }
+        return false;
+    }
+
+    @Override
+    public boolean visit(EnhancedForStatement enhancedFor) {
+        if (isKeySetMethod(enhancedFor.getExpression())) {
+            // From 'for (K key : map.keySet()) { }'
+            // -> mapExpression become 'map', parameter become 'K key'
+            final Expression mapExpression = ((MethodInvocation) enhancedFor.getExpression()).getExpression();
+            final SingleVariableDeclaration parameter = enhancedFor.getParameter();
+
+            final ITypeBinding valueType = findAndRefactor(mapExpression, parameter, enhancedFor.getBody());
+            if (valueType != null) {
+
+                final ASTBuilder b = ctx.getASTBuilder();
+                final AST ast = ctx.getAST();
+                final EnhancedForStatement newFor = ast.newEnhancedForStatement();
+                newFor.setExpression(b.invoke(b.copy(mapExpression), "entrySet"));
+                newFor.setParameter(b.declareSingleVariable(b.simpleName("entry"), "Map.Entry",
+                        getFriendlyTypeName(parameter.getType().resolveBinding()), getFriendlyTypeName(valueType)));
+
+                final Statement newBody = b.copy(enhancedFor.getBody());
+
+                // Declare 'KeyType' 'oldLoopVariable' = entry.getKey();
+                final VariableDeclarationStatement newKeyDeclaration = b.declareStmt(
+                        getFriendlyTypeName(parameter.getType().resolveBinding()),
```

Comment:
I do not think it is necessary to use a friendly type name here.
We should just use the type as it currently is in the code.